addons/crm/models/crm_lead.py

```diff
@@ -392,12 +410,28 @@ def _compute_contact_name(self):
         """ compute the new values when partner_id has changed """
         for lead in self:
             lead.update(lead._prepare_contact_name_from_partner(lead.partner_id))
+            
+    @api.depends('company_partner_id', 'contact_partner_id')
+    def _compute_partner_id(self):
+        for lead in self:
+            lead.partner_id = lead.contact_partner_id or lead.company_partner_id
+            
+    def _inverse_partner_id(self):
+        for lead in self.filtered('partner_id'):
+            if lead.partner_id.is_company:
+                lead.company_partner_id = lead.partner_id
+                lead.contact_partner_id = False
```

Comment:
In the corner case contact_partner_id is a company, we actually 'jump a generation' and end up setting the contact to false and use the child company instead. Corner case though.